pkg/lib/cockpit-connect-ssh.tsx

```diff
@@ -0,0 +1,639 @@
+ * target host).
+ *
+ * For all other cases, call this function first. It shows various dialogs where
+ * the user can specify a login password or unlock their SSH key. If the user
+ * does not already have an SSH key, the dialog also offers to create one. It
+ * also offers to authorize the user's SSH key to the remote machine/user.
+ *
+ * Arguments:
+ * @dialog_context: The page's `DialogsContext`, see ./dialogs.tsx
+ * @host: Same `[user@]host[:port]` format as the channel option; must be
+ *        *exactly* the same as for opening the channel afterwards
+ * @user: Same as the channel option; overrides `user@` portion of @host
+ * Returns: Nothing on success. Afterwards the SSH conection is established and
+ *          you can use the `host` option in channels. Throws a "cancel"
+ *          exception if the user cancelled the dialog. Most SSH errors are
+ *          handled in the dialogs, but you still have to expect and check for
+ *          other Cockpit errors with the usual `{ problem: "...", ... }` structure.
+ *
+ * See pkg/playground/remote.tsx for an example how to use this function.
+ *
+ * [1] https://github.com/cockpit-project/cockpit/blob/main/doc/protocol.md#command-init
+ */
+
+export async function connect_host(dialog_context: Dialogs, host: string, user?: string) {
+    const options: cockpit.ChannelOptions = { host };
+    if (user)
+        options.user = user;
+    while (true) {
+        try {
+            await try_connect(options);
+            debug(host, "succeeded");
+            break;
+        } catch (error) {
+            const ex = error as cockpit.JsonObject;
+            // unknown host or changed host key → re-try with private session to get its host key
+            if ((ex.problem === "unknown-host" || ex.problem === "invalid-hostkey") && !ex["host-key"]) {
+                debug(host, "failed with unknown-host, retrying with private session");
+                options.session = "private";
+                continue;
+            } else {
+                // reset
+                delete options.session;
+            }
+
+            // @ts-expect-error: ex is untyped, and this is too much useless hassle
+            const dialog = error_dialogs[ex.problem];
+            if (dialog) {
+                debug(host, "failed with:", ex, "mapping to", dialog);
+                try {
+                    const result = await dialog_context.run(dialog, { host, user, error: ex });
+                    debug(host, "dialog result:", result);
+                } catch (error) {
+                    const dialog_error = error as cockpit.JsonObject;
+                    debug(host, "dialog", dialog, "failed with:", dialog_error);
+                    if (dialog_error.problem === "no-cockpit") {
+                        // avoid another SSH connection
+                        await dialog_context.run(NotSupportedDialog, { host, user, error: dialog_error });
```

Comment:
This added line is not executed by any test.

Reply:
That's nonsense -- this *is* covered, if I remove the line, the test fails.